### PR TITLE
fix(replay): handle null user fields as a special case in search config

### LINKED ISSUE #78286: Inconsistent session replay search results
### Environment

SaaS (https://sentry.io/)

### Steps to Reproduce

1. navigate to Session Replay main page
2. search `user.email is not "" `
3. compare results of this search to the unfiltered results

### Expected Result

All of the replays with a user email appear on the list for the `user.email is not "" ` search

Image

### Actual Result

Some results that appeared in the unfiltered search (and had a user email) do not appear in the filtered search

Unfiltered search:
Image

Filtered search:
Image

### Product Area

Replays

### Link

_No response_

### DSN

_No response_

### Version

_No response_

### PR DESCRIPTION
**Pull request by aliu39**
Follow-up to https://github.com/getsentry/sentry/pull/78642. Closes https://github.com/getsentry/sentry/issues/78286

Because user.[ip, email, username, id] are nullable strings, and we aggregate the segments with `anyIf`, we can't use the row-by-row scalar config to filter on them. 

For the aggregate config, we need to handle the special case of null values. A replay's field is null if and only if all its segments have field = null.

If you try the [user.email:""](https://sentry.sentry.io/replays/?cursor=0%3A0%3A1&project=11276&query=user.email%3A%22%22&statsPeriod=7d&utc=true) filter in sentry right now, and paginate results, you can see many replays with an email are returned. We might even be returning all of them.

### CODE REVIEW

**Comment by aliu39** [comment]
STill need to fix for Mat View tests.

File: src/sentry/replays/usecases/query/conditions/aggregate.py [hunk 1]
```diff
@@ -61,65 +61,85 @@ def visit_not_in(expression: Expression, value: list[int]) -> Condition:
...
+        if value == "":
+            # "All rows in aggregation set are null."
+            return does_not_contain(StringScalar.visit_neq(expression, ""))
         return contains(StringScalar.visit_eq(expression, value))
 
     @staticmethod
     def visit_neq(expression: Expression, value: str) -> Condition:
+        if value == "":
+            # "1+ rows in aggregation set are non-null."
+            return contains(StringScalar.visit_neq(expression, ""))
```
Comment (cmanallen) [inline src/sentry/replays/usecases/query/conditions/aggregate.py:109]: This is hard to read.  Abstract `StringScalar.visit_neq(expression, "")` to a function called `filled(expr)` so you can write `contains(filled(expr))` and `does_not_contain(filled(expr))`.  Apply everywhere we're doing this empty check.

**Review by cmanallen — CHANGES_REQUESTED** [review]

File: tests/sentry/replays/test_organization_replay_index.py [hunk 2]
```diff
@@ -1717,63 +1717,81 @@ def test_query_invalid_ipv4_addresses(self):
...
                 data = response.json()["data"]
                 assert len(data) == 0
 
+    def test_query_null_email(self):
+        self._test_null_filters("user.email", "user_email", "", "andrew@example.com")
+
+    def test_query_null_ipv4(self):
+        self._test_null_filters("user.ip", "ipv4", None, "127.0.0.1")
+
+    def test_query_null_username(self):
```
[the remark this answers is not in the document]
Comment (aliu39) [reply tests/sentry/replays/test_organization_replay_index.py:1789]: For the test names, right?

File: src/sentry/replays/usecases/query/conditions/aggregate.py (hunk 1 above)
Comment (aliu39) [reply src/sentry/replays/usecases/query/conditions/aggregate.py:109]: The name `filled` is unclear to me. How about `nonempty` or `is_nonempty`?

Comment (cmanallen) [reply src/sentry/replays/usecases/query/conditions/aggregate.py:109]: Sure.

File: tests/sentry/replays/test_organization_replay_index.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (cmanallen) [reply tests/sentry/replays/test_organization_replay_index.py:1789]: Yes.

**Comment by aliu39** [comment]
@cmanallen Any ideas how to handle these cases for materialized view?

**Comment by cmanallen** [comment]
@aliu39 Don't worry about it.  If you're getting test failures just opt out of the materialized-view.  It needs to be removed.

**Review by cmanallen — CHANGES_REQUESTED** [review]
Do you mind adding coverage for these missed lines?  https://app.codecov.io/gh/getsentry/sentry/pull/79054?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=checks&utm_campaign=pr+comments&utm_term=getsentry

File: src/sentry/replays/usecases/query/conditions/aggregate.py [hunk 3]
```diff
@@ -40,6 +40,14 @@
 from sentry.replays.lib.new_query.utils import contains, does_not_contain
 
 
+def _nonempty_str(expression: Expression) -> Condition:
+    return StringScalar.visit_neq(expression, "")
+
+
+def _nonempty_ipv4(expression: Expression) -> Condition:
```
Comment (cmanallen) [inline src/sentry/replays/usecases/query/conditions/aggregate.py:47]: Nit: not_null would be a better prefix since it can be null.

**Review by cmanallen — APPROVED** [review]

**Pull request merged**